// license-check-and-add/dist/src/constants.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = __importStar(require("path"));
exports.OPTIONS = {
    'config-file': {
        alias: ['f'],
        default: path.join(process.cwd(), 'license-checker-config.json'),
        required: false,
        type: 'string',
    },
};
exports.CONFIG_OPTION = 'config-file';
// tslint:disable: object-literal-sort-keys
exports.DEFAULT_FORMAT = {
    prepend: '/*',
    append: '*/',
};
exports.DEFAULT_FORMATS = {
    'gitignore|npmignore|eslintignore|dockerignore|sh|py': {
        eachLine: {
            prepend: '# ',
        },
    },
    'html|xml|svg': {
        prepend: '<!--',
        append: '-->',
    },
    'js|ts|css|scss|less|php|as|c|java|cpp|go|cto|acl': {
        prepend: '/*',
        append: '*/',
    },
    txt: {},
};
// tslint:enable: object-literal-sort-keys
